File: MoonEngine-Root/ECS/src/test/java/com/greentree/common/ecs/TestUtil.java
package com.greentree.common.ecs;

import static org.junit.jupiter.api.Assertions.*;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;

import com.greentree.engine.moon.ecs.Entity;

public class TestUtil {

	public static void assertComponentEquals(Entity a, Entity b) {
		for(var ac : a) {
			final var bc = b.get(ac.getClass());
			assertEquals(ac, bc);
		}
		for(var bc : b) {
			final var ac = a.get(bc.getClass());
			assertEquals(ac, bc);
		}
	}


	@SuppressWarnings("unchecked")
	public static <T> T deser(byte[] ser) throws IOException, ClassNotFoundException {
		try(final var bout = new ByteArrayInputStream(ser)) {
			try(final var oout = new ObjectInputStream(bout)) {
				return (T) oout.readObject();
			}
		}
	}

	public static byte[] ser(Object obj) throws IOException {
		try(final var bout = new ByteArrayOutputStream()) {
			try(final var oout = new ObjectOutputStream(bout)) {
				oout.writeObject(obj);
			}
			return bout.toByteArray();
		}
	}
	
}

File: MoonEngine-Root/ECS/src/test/java/com/greentree/common/ecs/TestUtil.kt
package com.greentree.common.ecs

import com.greentree.engine.moon.ecs.Entity
import org.junit.jupiter.api.Assertions
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.ObjectInputStream
import java.io.ObjectOutputStream

object TestUtil {

	@JvmStatic
	fun assertComponentEquals(a: Entity, b: Entity) {
		for (ac in a) {
			val bc = b[ac.javaClass]
			Assertions.assertEquals(ac, bc)
		}
		for (bc in b) {
			val ac = a[bc.javaClass]
			Assertions.assertEquals(ac, bc)
		}
	}

	@JvmStatic
	fun <T> deser(ser: ByteArray?): T {
		ByteArrayInputStream(ser).use { bout -> ObjectInputStream(bout).use { oout -> return oout.readObject() as T } }
	}

	@JvmStatic
	fun ser(obj: Any): ByteArray {
		ByteArrayOutputStream().use { bout ->
			ObjectOutputStream(bout).use { oout -> oout.writeObject(obj) }
			return bout.toByteArray()
		}
	}
}